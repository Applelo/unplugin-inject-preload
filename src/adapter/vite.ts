import type { HtmlTagDescriptor, Logger } from 'vite'
import type { OutputBundle } from 'rollup'
import { getAssetsForViteJS } from '../helper/getAssets'
import type { Options } from '../types'
import { getTagsAttributes } from '../helper/getTagsAttributes'
import { serializeTags } from '../helper/serializer'

export function viteAdapter(args: {
  bundle: OutputBundle
  html: string
  options: Options
  customInject: RegExp
  viteBasePath: string
  viteLogger: Logger
}) {
  const {
    bundle,
    html,
    options,
    customInject,
    viteBasePath,
    viteLogger,
  } = args

  const injectTo
          = (options.injectTo && options.injectTo !== 'custom')
            ? options.injectTo
            : 'head-prepend'

  const assets = getAssetsForViteJS(bundle)
  const tags: HtmlTagDescriptor[] = []
  const tagsAttributes = getTagsAttributes(
    assets,
    options,
    viteBasePath,
    viteLogger,
  )

  tagsAttributes.forEach((attrs) => {
    tags.push({
      tag: 'link',
      attrs,
      injectTo,
    })
  })

  if (options.injectTo === 'custom') {
    return html.replace(
      customInject,
      (match, p1) => `\n${serializeTags(tags, p1)}`,
    )
  }
  else {
    return tags
  }
}
