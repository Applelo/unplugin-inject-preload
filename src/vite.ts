/**
 * This entry file is for Vite plugin.
 *
 * @module
 */

import unplugin from './index'

/**
 * Vite plugin
 *
 * @example
 * ```ts
 * // vite.config.ts
 * import UnpluginInjectPreload from 'unplugin-inject-preload/vite'
 *
 * export default defineConfig({
 *   plugins: [
 *     UnpluginInjectPreload({ your_options }),
 *   ],
 * })
 * ```
 */
export default unplugin.vite
