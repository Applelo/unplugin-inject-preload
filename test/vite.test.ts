import { resolve } from 'node:path'
import type { OutputAsset, RollupOutput } from 'rollup'
import { type InlineConfig, build } from 'vite'
import { describe, expect, it } from 'vitest'
import UnpluginInjectPreload from './../src/vite'
import type { Options } from './../src/types'

const configs: Record<string, Options> = {
  injectBottom: {
    files: [
      {
        match: /Roboto-[a-zA-Z]*-[a-z-0-9]*\.woff2$/,
      },
      {
        match: /lazy.[a-z-0-9]*.(css|js)$/,
      },
    ],
    injectTo: 'head',
  },
  customAttributes: {
    files: [
      {
        match: /Roboto-[a-zA-Z]*-[a-z-0-9]*\.woff2$/,
        attributes: {
          'as': 'font',
          'crossorigin': 'anonymous',
          'data-font': 'Roboto',
          'type': 'font/woff2',
        },
      },
    ],
  },
  auto: {
    files: [
      {
        match: /Roboto-[a-zA-Z]*-[a-z-0-9]*\.woff2$/,
      },
      {
        match: /lazy.[a-z-0-9]*.(css|js)$/,
      },
    ],
  },
  customPosition: {
    files: [
      {
        match: /Roboto-[a-zA-Z]*-[a-z-0-9]*\.woff2$/,
        attributes: {
          'data-vite-plugin-inject-preload': true,
        },
      },
      {
        match: /lazy.[a-z-0-9]*.(css|js)$/,
      },
    ],
    injectTo: 'custom',
  },
  wrongAttributes: {
    files: [
      {
        match: /Roboto-[a-zA-Z]*-[a-z-0-9]*\.woff2$/,
        attributes: {
          href: './yolo.woff2',
        },
      },
    ],
  },
  noType: {
    files: [
      {
        match: /Roboto-[a-zA-Z]*-[a-z-0-9]*\.woff2$/,
        attributes: {
          type: undefined,
        },
      },
    ],
  },
  modulepreload: {
    files: [
      {
        match: /lazy.[a-z-0-9]*.(js)$/,
        attributes: {
          rel: 'modulepreload',
        },
      },
    ],
  },
}

const buildVite = async (pluginConfig: Options, config: InlineConfig = {}) => {
  const { output } = (await build({
    root: resolve(__dirname, './fixtures/vitejs'),
    plugins: [UnpluginInjectPreload(pluginConfig)],
    ...config,
  })) as RollupOutput

  const { source: indexSource } = output.find(
    item => item.fileName === 'index.html',
  ) as OutputAsset

  return indexSource.toString()
}

describe('excerpt vitejs', () => {
  for (const key in configs) {
    if (Object.prototype.hasOwnProperty.call(configs, key)) {
      const config = configs[key] as Options
      it(`test ${key}`, async () => {
        const output = await buildVite(config)
        expect(output).toMatchSnapshot()
      })

      it(`test ${key} with basePath`, async () => {
        const output = await buildVite(config, { base: '/base' })
        expect(output).toMatchSnapshot()
      })
    }
  }
})
