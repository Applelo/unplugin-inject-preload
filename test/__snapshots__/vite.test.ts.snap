// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expect vitejs > test auto 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <link
      rel="preload"
      href="/assets/Roboto-Italic-C4aLNNPR.woff2"
      type="font/woff2"
      as="font"
    />
    <link
      rel="preload"
      href="/assets/Roboto-Regular-ZlapIrEx.woff2"
      type="font/woff2"
      as="font"
    />
    <link
      rel="preload"
      href="/assets/lazy-BMgJfzmm.js"
      type="application/javascript"
      as="script"
    />

    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script type="module" crossorigin src="/assets/index-Dbf-6TCm.js"></script>
    <link rel="stylesheet" crossorigin href="/assets/index-BkR2O4Mb.css" />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test auto with basePath 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <link
      rel="preload"
      href="/base/assets/Roboto-Italic-C4aLNNPR.woff2"
      type="font/woff2"
      as="font"
    />
    <link
      rel="preload"
      href="/base/assets/Roboto-Regular-ZlapIrEx.woff2"
      type="font/woff2"
      as="font"
    />
    <link
      rel="preload"
      href="/base/assets/lazy-BMgJfzmm.js"
      type="application/javascript"
      as="script"
    />

    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script
      type="module"
      crossorigin
      src="/base/assets/index-C0NrOB6i.js"
    ></script>
    <link rel="stylesheet" crossorigin href="/base/assets/index-BFryvKQb.css" />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test customAttributes 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <link
      rel="preload"
      href="/assets/Roboto-Italic-C4aLNNPR.woff2"
      type="font/woff2"
      as="font"
      crossorigin="anonymous"
      data-font="Roboto"
    />
    <link
      rel="preload"
      href="/assets/Roboto-Regular-ZlapIrEx.woff2"
      type="font/woff2"
      as="font"
      crossorigin="anonymous"
      data-font="Roboto"
    />

    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script type="module" crossorigin src="/assets/index-Dbf-6TCm.js"></script>
    <link rel="stylesheet" crossorigin href="/assets/index-BkR2O4Mb.css" />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test customAttributes with basePath 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <link
      rel="preload"
      href="/base/assets/Roboto-Italic-C4aLNNPR.woff2"
      type="font/woff2"
      as="font"
      crossorigin="anonymous"
      data-font="Roboto"
    />
    <link
      rel="preload"
      href="/base/assets/Roboto-Regular-ZlapIrEx.woff2"
      type="font/woff2"
      as="font"
      crossorigin="anonymous"
      data-font="Roboto"
    />

    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script
      type="module"
      crossorigin
      src="/base/assets/index-C0NrOB6i.js"
    ></script>
    <link rel="stylesheet" crossorigin href="/base/assets/index-BFryvKQb.css" />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test customPosition 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>

    <link
      rel="preload"
      href="/assets/Roboto-Italic-C4aLNNPR.woff2"
      type="font/woff2"
      as="font"
      data-vite-plugin-inject-preload
    />
    <link
      rel="preload"
      href="/assets/Roboto-Regular-ZlapIrEx.woff2"
      type="font/woff2"
      as="font"
      data-vite-plugin-inject-preload
    />
    <link
      rel="preload"
      href="/assets/lazy-BMgJfzmm.js"
      type="application/javascript"
      as="script"
    />

    <script type="module" crossorigin src="/assets/index-Dbf-6TCm.js"></script>
    <link rel="stylesheet" crossorigin href="/assets/index-BkR2O4Mb.css" />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test customPosition with basePath 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>

    <link
      rel="preload"
      href="/base/assets/Roboto-Italic-C4aLNNPR.woff2"
      type="font/woff2"
      as="font"
      data-vite-plugin-inject-preload
    />
    <link
      rel="preload"
      href="/base/assets/Roboto-Regular-ZlapIrEx.woff2"
      type="font/woff2"
      as="font"
      data-vite-plugin-inject-preload
    />
    <link
      rel="preload"
      href="/base/assets/lazy-BMgJfzmm.js"
      type="application/javascript"
      as="script"
    />

    <script
      type="module"
      crossorigin
      src="/base/assets/index-C0NrOB6i.js"
    ></script>
    <link rel="stylesheet" crossorigin href="/base/assets/index-BFryvKQb.css" />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test entryMatch 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script type="module" crossorigin src="/assets/index-Dbf-6TCm.js"></script>
    <link rel="stylesheet" crossorigin href="/assets/index-BkR2O4Mb.css" />
    <link
      rel="preload"
      href="/assets/Roboto-Italic-C4aLNNPR.woff2"
      type="font/woff2"
      as="font"
    />
    <link
      rel="preload"
      href="/assets/Roboto-Regular-ZlapIrEx.woff2"
      type="font/woff2"
      as="font"
    />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test entryMatch with basePath 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script
      type="module"
      crossorigin
      src="/base/assets/index-C0NrOB6i.js"
    ></script>
    <link rel="stylesheet" crossorigin href="/base/assets/index-BFryvKQb.css" />
    <link
      rel="preload"
      href="/base/assets/Roboto-Italic-C4aLNNPR.woff2"
      type="font/woff2"
      as="font"
    />
    <link
      rel="preload"
      href="/base/assets/Roboto-Regular-ZlapIrEx.woff2"
      type="font/woff2"
      as="font"
    />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test injectBottom 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script type="module" crossorigin src="/assets/index-Dbf-6TCm.js"></script>
    <link rel="stylesheet" crossorigin href="/assets/index-BkR2O4Mb.css" />
    <link
      rel="preload"
      href="/assets/Roboto-Italic-C4aLNNPR.woff2"
      type="font/woff2"
      as="font"
    />
    <link
      rel="preload"
      href="/assets/Roboto-Regular-ZlapIrEx.woff2"
      type="font/woff2"
      as="font"
    />
    <link
      rel="preload"
      href="/assets/lazy-BMgJfzmm.js"
      type="application/javascript"
      as="script"
    />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test injectBottom with basePath 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script
      type="module"
      crossorigin
      src="/base/assets/index-C0NrOB6i.js"
    ></script>
    <link rel="stylesheet" crossorigin href="/base/assets/index-BFryvKQb.css" />
    <link
      rel="preload"
      href="/base/assets/Roboto-Italic-C4aLNNPR.woff2"
      type="font/woff2"
      as="font"
    />
    <link
      rel="preload"
      href="/base/assets/Roboto-Regular-ZlapIrEx.woff2"
      type="font/woff2"
      as="font"
    />
    <link
      rel="preload"
      href="/base/assets/lazy-BMgJfzmm.js"
      type="application/javascript"
      as="script"
    />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test modulepreload 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <link
      rel="modulepreload"
      href="/assets/lazy-BMgJfzmm.js"
      type="application/javascript"
      as="script"
    />

    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script type="module" crossorigin src="/assets/index-Dbf-6TCm.js"></script>
    <link rel="stylesheet" crossorigin href="/assets/index-BkR2O4Mb.css" />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test modulepreload with basePath 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <link
      rel="modulepreload"
      href="/base/assets/lazy-BMgJfzmm.js"
      type="application/javascript"
      as="script"
    />

    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script
      type="module"
      crossorigin
      src="/base/assets/index-C0NrOB6i.js"
    ></script>
    <link rel="stylesheet" crossorigin href="/base/assets/index-BFryvKQb.css" />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test noMatch 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script type="module" crossorigin src="/assets/index-Dbf-6TCm.js"></script>
    <link rel="stylesheet" crossorigin href="/assets/index-BkR2O4Mb.css" />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test noMatch with basePath 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script
      type="module"
      crossorigin
      src="/base/assets/index-C0NrOB6i.js"
    ></script>
    <link rel="stylesheet" crossorigin href="/base/assets/index-BFryvKQb.css" />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test noType 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <link rel="preload" href="/assets/Roboto-Italic-C4aLNNPR.woff2" as="font" />
    <link
      rel="preload"
      href="/assets/Roboto-Regular-ZlapIrEx.woff2"
      as="font"
    />

    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script type="module" crossorigin src="/assets/index-Dbf-6TCm.js"></script>
    <link rel="stylesheet" crossorigin href="/assets/index-BkR2O4Mb.css" />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test noType with basePath 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <link
      rel="preload"
      href="/base/assets/Roboto-Italic-C4aLNNPR.woff2"
      as="font"
    />
    <link
      rel="preload"
      href="/base/assets/Roboto-Regular-ZlapIrEx.woff2"
      as="font"
    />

    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script
      type="module"
      crossorigin
      src="/base/assets/index-C0NrOB6i.js"
    ></script>
    <link rel="stylesheet" crossorigin href="/base/assets/index-BFryvKQb.css" />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test wrongAttributes 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <link rel="preload" href="./yolo.woff2" type="font/woff2" as="font" />
    <link rel="preload" href="./yolo.woff2" type="font/woff2" as="font" />

    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script type="module" crossorigin src="/assets/index-Dbf-6TCm.js"></script>
    <link rel="stylesheet" crossorigin href="/assets/index-BkR2O4Mb.css" />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;

exports[`expect vitejs > test wrongAttributes with basePath 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <link rel="preload" href="./yolo.woff2" type="font/woff2" as="font" />
    <link rel="preload" href="./yolo.woff2" type="font/woff2" as="font" />

    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Unplugin Inject Preload</title>
    <!--__unplugin-inject-preload__-->
    <script
      type="module"
      crossorigin
      src="/base/assets/index-C0NrOB6i.js"
    ></script>
    <link rel="stylesheet" crossorigin href="/base/assets/index-BFryvKQb.css" />
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
  </body>
</html>
"
`;
